slyd/slyd/orm/models.py

```diff
@@ -0,0 +1,585 @@
+    vary = Boolean(default=False)
+    auto_created = Boolean(default=False)
+    schema = BelongsTo(Schema, related_name='fields', on_delete=CASCADE,
+                       ignore_in_file=True)
+    annotations = HasMany('Annotation', related_name='field',
+                          on_delete=CASCADE_AUTO_OR_CLEAR, ignore_in_file=True)
+
+    class Meta:
+        path = u'items.json'
+        owner = 'schema'
+        envelope = True
+
+    def __repr__(self):
+        return super(Field, self).__repr__('name', 'type')
+
+    @pre_load
+    def name_from_id(self, data):
+        if 'name' not in data:
+            data['name'] = data['id']
+        return data
+
+    @pre_load
+    def add_fake_annotations(self, data):
+        data['annotations'] = []
+        return data
+
+    @post_dump
+    def remove_fake_annotations(self, data):
+        data.pop('annotations', None)
+        return data
+
+    @post_dump
+    def remove_auto_created_false(self, data):
+        if 'auto_created' in data and not data['auto_created']:
+            del data['auto_created']
+        return data
+
+
+class Extractor(Model):
+    id = String(primary_key=True)
+    type = String(required=True, validate=OneOf(['type', 'regex']))
+    value = DependantField(when='type', then={
+        'type': String(required=True, validate=OneOf(FIELD_TYPES)),
+        'regex': Regexp(required=True),
+    })
+    project = BelongsTo(Project, related_name='extractors', on_delete=CASCADE,
+                        ignore_in_file=True)
+    annotations = HasMany('Annotation', related_name='extractors',
+                          on_delete=CLEAR, ignore_in_file=True)
+
+    class Meta:
+        path = u'extractors.json'
+        owner = 'project'
+        envelope = True
+
+    @pre_load
+    def to_type_and_value(self, data):
+        type_extractor = data.pop('type_extractor', None)
+        regular_expression = data.pop('regular_expression', None)
+        if type_extractor:
+            data['type'] = 'type'
+            data['value'] = type_extractor
+        elif regular_expression:
+            data['type'] = 'regex'
+            data['value'] = regular_expression
+        return data
+
+    @post_dump
+    def from_type_and_value(self, data):
+        # may be missing if serializing only id
+        if 'type' in data and 'value' in data:
+            type_ = data.pop('type')
+            value = data.pop('value')
+            if type_ == 'type':
+                data['type_extractor'] = value
+            else:  # type_ == 'regex'
+                data['regular_expression'] = value
+        return data
+
+
+class Spider(Model):
+    # TODO: validate id against allowed file name
+    id = String(primary_key=True)
+    start_urls = List(Url)
+    # TODO: generated urls
+    links_to_follow = String(default='all', validate=OneOf(
+        ['none', 'patterns', 'all', 'auto']))
+    # TODO: compute automatically from start urls
+    allowed_domains = List(Domain)
```

Comment:
allowed_domains is already generated from start_urls (and sample urls) when a spider is run. This feature could be useful if a user wants to define custom allowed domains.